cdc/owner/scheduler.go

```diff
@@ -0,0 +1,358 @@
+		moveTableTargets: make(map[model.TableID]model.CaptureID),
+	}
+}
+
+// Tick is a main process of scheduler, it dispatches tables into captures, and handles move-table and rebalance events.
+// Tick returns a bool represents the changefeed state can be update in this tick.
+// The state can be update only if all the tables which should be listened dispatched to captures and no operation sent to captures in this tick.
+func (s *scheduler) Tick(state *model.ChangefeedReactorState, currentTables []model.TableID, captures map[model.CaptureID]*model.CaptureInfo) (shouldUpdateState bool) {
+	s.state = state
+	s.currentTables = currentTables
+	s.captures = captures
+
+	s.cleanUpFinishedOperations()
+	pendingJob := s.syncTablesWithSchemaManager()
+	s.dispatchToTargetCaptures(pendingJob)
+	if len(pendingJob) != 0 {
+		log.Debug("scheduler:generated pending job to be executed", zap.Any("pendingJob", pendingJob))
+	}
+	s.handleJobs(pendingJob)
+	shouldUpdateState = len(pendingJob) == 0
+	shouldUpdateState = s.rebalance() && shouldUpdateState
+	shouldUpdateState = s.handleMoveTableJob() && shouldUpdateState
+	s.lastTickCaptureCount = len(captures)
+	return shouldUpdateState
+}
+
+func (s *scheduler) MoveTable(tableID model.TableID, target model.CaptureID) {
+	s.moveTableJobQueue = append(s.moveTableJobQueue, &moveTableJob{
+		tableID: tableID,
+		target:  target,
+	})
+}
+
+func (s *scheduler) handleMoveTableJob() (shouldUpdateState bool) {
+	shouldUpdateState = true
+	if len(s.moveTableJobQueue) == 0 {
+		return
+	}
+	table2CaptureIndex := s.table2CaptureIndex()
+	for _, job := range s.moveTableJobQueue {
+		source, exist := table2CaptureIndex[job.tableID]
+		if !exist {
+			return
+		}
+		s.moveTableTargets[job.tableID] = job.target
+		job := job
+		shouldUpdateState = false
+		s.state.PatchTaskStatus(source, func(status *model.TaskStatus) (*model.TaskStatus, bool, error) {
+			if status == nil {
+				// the capture may be down, just skip remove this table
+				return status, false, nil
+			}
+			if status.Operation != nil && status.Operation[job.tableID] != nil {
+				// skip removing this table to avoid the remove operation created by the rebalance function interfering with the operation created by another function
+				return status, false, nil
+			}
+			status.RemoveTable(job.tableID, s.state.Status.CheckpointTs, false)
+			return status, true, nil
+		})
+	}
+	s.moveTableJobQueue = nil
+	return
+}
+
+func (s *scheduler) Rebalance() {
+	s.needRebalanceNextTick = true
+}
+
+func (s *scheduler) table2CaptureIndex() map[model.TableID]model.CaptureID {
+	table2CaptureIndex := make(map[model.TableID]model.CaptureID)
+	for captureID, taskStatus := range s.state.TaskStatuses {
+		for tableID := range taskStatus.Tables {
+			if preCaptureID, exist := table2CaptureIndex[tableID]; exist && preCaptureID != captureID {
+				log.Panic("A table is replicated by two processor, please report a bug", zap.Strings("capture-ids", []string{preCaptureID, captureID}))
```

Comment:
IMHO return an error and resign the owner here is better than crashes the cdc server. Either panic or return error will trigger a new owner election, but panic will also let the processors on this node to be rebalanced.